src/webapi/indexeddb.rs

```diff
@@ -0,0 +1,1145 @@
+    DataError,
+    ///	The cursor is currently being iterated or has iterated past its end.
+    InvalidStateError,
+    ///	The cursor's direction is not prev or next.
+    InvalidAccessError
+}
+
+error_enum_boilerplate! {
+    IDBUpdateError,
+    /// This IDBCursor's transaction is inactive.
+    TransactionInactiveError,
+    /// The transaction mode is read only.
+    ReadOnlyError,
+    /// The cursor was created using IDBIndex.openKeyCursor, is currently being iterated, or has iterated past its end.
+    InvalidStateError,
+    /// The underlying object store uses in-line keys and the property in the value at the object store's key path does not match the key in this cursor's position.
+    DataError,
+    ///The data being stored could not be cloned by the internal structured cloning algorithm.
+    DataCloneError
+}
+
+///
+error_enum_boilerplate! {
+    ///
+    IDBAddError,
+    /// This IDBCursor's transaction is inactive.
+    TransactionInactiveError,
+    /// The transaction mode is read only.
+    ReadOnlyError,
+    /// The cursor was created using IDBIndex.openKeyCursor, is currently being iterated, or has iterated past its end.
+    InvalidStateError,
+    /// The underlying object store uses in-line keys and the property in the value at the object store's key path does not match the key in this cursor's position.
+    DataError,
+    ///The data being stored could not be cloned by the internal structured cloning algorithm.
+    DataCloneError,
+    /// An operation failed because the primary key constraint was violated (due to an already existing record with the same primary key value).
+    ConstraintError
+}
+
+error_enum_boilerplate! {
+    IDBCursorDeleteError,
+    /// This IDBCursor's transaction is inactive.
+    TransactionInactiveError,
+    /// The transaction mode is read-only.
+    ReadOnlyError,
+    /// The cursor was created using IDBindex.openKeyCursor, is currently being iterated, or has iterated past its end.
+    InvalidStateError
+}
+
+error_enum_boilerplate! {
+    IDBClearError,
+    /// The transaction associated with this operation is in read-only mode.
+    ReadOnlyError,
+    /// This IDBObjectStore's transaction is inactive.
+    TransactionInactiveError
+}
+
+/// This trait implements all the methods that are shared between
+/// `IDBCursor` and `IDBCursorWithValue`.
+pub trait IDBCursorSharedMethods: AsRef< Reference >  {
+    
+    /// The source read-only property of the `IDBCursor` interface returns
+    /// the `IDBObjectStore` or `IDBIndex` that the cursor is iterating over.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/IDBCursor/source)
+    fn source( &self ) -> IDBCursorSource {
+        if js!( return @{self.as_ref()}.source instanceof IDBObjectStore; ).try_into().unwrap() {
+            IDBCursorSource::Store(js!( return @{self.as_ref()}.source ).try_into().unwrap())
```

Comment:
1. You can do `let source = js!( return @{self.as_ref()}.source; );` only once at the start of the function.
2. Missing semicolon in the JS snippet.